src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const historyClones = [];
+  const currentSt = { ...state };
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'clear':
+        for (const i in currentSt) {
+          delete currentSt[i];
+        }
+        break;
+
+      case 'addProperties':
+        Object.assign(currentSt, extraData);
+        break;
+
+      case 'removeProperties':
+        for (const i of keysToRemove) {
+          if (currentSt[i]) {
+            delete currentSt[i];
+          }
+        }
+        break;
+      default:
+        throw new Error(`unexpected action ${action}`);
+    }
+
+    historyClones.push({ ...currentSt });
```

Comment:
You are pushing a new state copy to 'historyClones' after each action, which is correct. However, to adhere to the 'Don't Repeat Yourself' principle, this operation should be performed once after the switch statement, not in each case.